research.rst

```diff
@@ -0,0 +1,315 @@
+just as we currently do for margarita.  Installing tequila would make
+available the ``tequila`` command, providing a simplified interface
+for ``ansible-playbook`` (and friends) that will be mostly familiar to
+current users of the django-project-template fabric scripts.  There
+would also be some mild changes to the project structure, i.e. the
+location of the project-specific variables and custom playbooks, since
+these would be kept distinct from paths where they might conflict with
+the equivalent Salt files.
+
+Much of this work has already been started, and is currently on the
+``use_tequila`` branch.
+
+
+Installation
+------------
+
+We need to have the ability to pin projects to particular versions of
+tequila.  It is also desirable to have an easy command to install or
+upgrade tequila, along the lines of ``pip install``.
+
+Under all of the proposals that follow, I recommend that Tequila be
+pip installable.  The work for this has already been done.
+
+
+Proposal 1: ``tequila roles`` command puts the roles in the right place
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Under this proposal, no changes to the structure of the Tequila
+repository need to happen.  Running the ``tequila roles`` command
+would symlink or copy the ``tequila/roles`` directory up to the
+current working directory (which under most scenarios would be the
+project directory), and this would generally not need to be run more
+than once per local setup.
+
+For development of the roles themselves, one could ``pip install -e
+<local_tequila_copy>``, and the in-progress versions of the roles
+would easily be made available for the ansible commands.
+
+The pros:
+
+- no changes to our Tequila repo structure
+- could still run plain versions of ``ansible-playbook``
+
+The cons:
+
+- our roles would not be easily usable by members of the Ansible
+  community
+
+
+Proposal 2: separate ``tequila-<rolename>`` repos
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If we retain Tequila as an open-sourced project, we should conform to
+the conventions of the Ansible community to the best of our ability.
+Since the Ansible community uses Ansible Galaxy (think: PyPI) and the
+``ansible-galaxy`` command (think: pip) for packaging and distribution
+of roles, we should do so as well.
+
+So, what do we need to do to make tequila installable using
+``ansible-galaxy``?  Roles are limited to one per repo or package,
+with all of the directory structure for the role exposed at the top
+level, i.e.
+
+::
+
+    myrole-repo/
+        defaults/
+        files/
+        handlers/
+        meta/
+        tasks/
+        templates/
+        tests/
+        vars/
+
+
+Tequila, as it currently stands, both fails to have the structure
+necessary for the installation process to complete without errors, and
+also has additional layers of directories, preventing roles from being
+correctly included from within playbooks.
```

Comment:
Actually, the wrapper program sets up the roles path so Ansible can find the roles, so role inclusion should work fine.